design/ratelimit-design-new.md

```diff
@@ -0,0 +1,180 @@
+| Global | [envoy.filters.network.ratelimit](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/rate_limit_filter#config-network-filters-rate-limit) | [envoy.filters.http.ratelimit](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_filter#config-http-filters-rate-limit) |
+
+This document focuses on the HTTP filters, as it's assumed that users will want L7 control over rate limiting.
+
+## Goals
+- Support Envoy's L7 **local** rate limiting filter.
+- Support Envoy's L7 **global** rate limiting filter, with a "bring your own Rate Limit Service (RLS)" model.
+
+## Non Goals
+- L4 rate limiting support (local or global).
+- Tight integration with a particular RLS.
+
+## High-Level Design
+
+A new type, `RateLimitPolicy`, will be defined as part of the `HTTPProxy` API.
+A `RateLimitPolicy` can be defined for either a virtual host or a route.
+The `RateLimitPolicy` defines parameters for local and/or global rate limiting.
+
+
+For local rate limiting, the user defines the rate limit itself, via [token bucket](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto#envoy-v3-api-field-extensions-filters-http-local-ratelimit-v3-localratelimit-token-bucket) settings.
+
+
+For global rate limiting, the user defines the descriptors to be generated and sent to the external RLS.
+Descriptors contain entries including things like: the client IP, the value of a particular header, the destination cluster, etc.
+The external RLS makes the rate limit decision based on the descriptors, and returns either a 200 or a 429 to Envoy.
+The operator of the external RLS must configure it with actual rate limits for different descriptors.
+Each RLS implementation may have its own configuration format and mechanism for defining rate limits for descriptors, so Contour cannot provide a generic API for defining these.
+
+
+For global rate limiting, the details of the external RLS are configured via an `ExtensionService` and in the Contour config file.
+Note that the main reason this is defined globally for Contour in the config file, rather than per-virtual host, is that there's a rate limit filter per HTTP connection manager, and all non-TLS virtual hosts share a single HTTP connection manager, so they necessarily must share the same rate limit filter config.
+If no global RLS is defined, no global rate limiting will occur.
+
+## Detailed Design
+
+### RateLimitPolicy type
+
+A sample `RateLimitPolicy` looks like:
+
+```yaml
+rateLimitPolicy:
+  # local defines local rate limiting properties for the virtual host or route.
+  local:
+    # tokenBucket defines the properties of the local rate limiting token bucket.
+    # ref. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket
+    tokenBucket:
+      maxTokens: 100
+      tokensPerFill: 10
+      fillInterval: 1s
```

Comment:
I'm not that comfortable exposing token bucket parameters directly; that's not how people who want to rate limit think about their problem. Could we re-cast this in terms of requests per second, burst, etc?

Reply:
Yeah, we can do that. It does mean that there's not a clear mapping to the Envoy config anymore, so anyone who wants to dig into how it translates has to do some mental gymnastics, but I guess that's Contour's MO.

Reply:
I think the main challenge here is to decide on a fill interval - we could arbitrarily set it to `1s` or some other fixed value. qps and burst don't really allow a user to express what they want there. I suppose, if we let the user specify both the "requests per time period" (e.g. 100) and the "time period duration" (e.g. 1 hour), then we could set fill interval equal to the "time period duration".

Reply:
TBQH I find token buckets a lot easier to reason about, but maybe that's just me :)

Reply:
>TBQH I find token buckets a lot easier to reason about, but maybe that's just me :)

That surprises me. I would have thought that people think about rate limiting as rates (ie. events per unit time). 

* The Gloo API is requests per time unit https://docs.solo.io/gloo-edge/latest/guides/security/rate_limiting/crds/
* The Ambassador API is rate per unit https://www.getambassador.io/docs/latest/howtos/advanced-rate-limiting/
* ingress nginx allows request/sec and request/min https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#rate-limiting